### ISSUE TITLE
KeyError on i3.metal instance types

### ISSUE THREAD

**Issue by barrymccallarcadia** [yahoo/ariel#3]
## Description
Spitting a stack trace: 
```'i3.metal': KeyError
Traceback (most recent call last):
  File "/var/task/ariel/lambda.py", line 106, in handler
    lambda_main(utils.load_config(config))
  File "/var/task/ariel/lambda.py", line 42, in lambda_main
    reports = generate_reports.generate(config, instances, ris, pricing)
  File "/var/task/ariel/generate_reports.py", line 52, in generate
    units_value = instances['instancetype'].apply(get_units) * instances['instances']
  File "/var/task/pandas/core/series.py", line 3591, in apply
    mapped = lib.map_infer(values, f, convert=convert_dtype)
  File "pandas/_libs/lib.pyx", line 2217, in pandas._libs.lib.map_infer
  File "/var/task/ariel/generate_reports.py", line 25, in get_units
    raise e
  File "/var/task/ariel/generate_reports.py", line 22, in get_units
    return pricing['units'][instancetype]
KeyError: 'i3.metal'
-----------------------------------------------------------
{
  "errorMessage": "'i3.metal'",
  "errorType": "KeyError",
  "stackTrace": [
    [
      "/var/task/ariel/lambda.py",
      106,
      "handler",
      "lambda_main(utils.load_config(config))"
    ],
    [
      "/var/task/ariel/lambda.py",
      42,
      "lambda_main",
      "reports = generate_reports.generate(config, instances, ris, pricing)"
    ],
    [
      "/var/task/ariel/generate_reports.py",
      52,
      "generate",
      "units_value = instances['instancetype'].apply(get_units) * instances['instances']"
    ],
    [
      "/var/task/pandas/core/series.py",
      3591,
      "apply",
      "mapped = lib.map_infer(values, f, convert=convert_dtype)"
    ],
    [
      "pandas/_libs/lib.pyx",
      2217,
      "pandas._libs.lib.map_infer",
      ""
    ],
    [
      "/var/task/ariel/generate_reports.py",
      25,
      "get_units",
      "raise e"
    ],
    [
      "/var/task/ariel/generate_reports.py",
      22,
      "get_units",
      "return pricing['units'][instancetype]"
    ]
  ]
}
```
## Expected Behavior
Not hotdog... I mean not stack trace. 

## Actual Behavior
~

## Possible Fix
¯\\\_(ツ)_/¯  not sure if i3.metal instances are just a special snowflake because they're bare metal.

## Steps to Reproduce
Have some running i3.metal instances and try to run

## Your Environment
* Version used: @meckstmd 2.0.1 version fix for issue #1

**Comment by meckstmd** [comment]
@barrymccallarcadia I'm glad you got farther by using the 2.0.1 version.  I looked through the code some today and it retrieves ec2 pricing data from an Amazon URL but then only loads the actual pricing data based on what Athena finds it has (based on instance and region type) from its data load based on the time you are calculating.  As best I can tell from a guess based on that error, it is not loading the `i3.metal` in its load of the pricing data for the region your instance is in but somehow you have an instance which is an i3.metal.  However, there is also code limiting instances to have a `Shared` product tenancy which a metal instance should not have.  I will need to run this and verify if I can recreate it and go from there.  Please send me what region(s) your i3.metal instances are in.  Also, can you take a look at your CloudWatch logs (you will have to change your config to use a `LOG_LEVEL` of `INFO`.  Even though it says it is the default, you need to specify it and not just have nothing there) and send me your logs. It should look something like this:
```
[INFO]	2019-07-22T13:00:43.277Z	8f9c695d-a5a1-4e22-838c-d40a8178d7a8	Loading Account Names...
[INFO]	2019-07-22T13:00:55.801Z	8f9c695d-a5a1-4e22-838c-d40a8178d7a8	Loaded 1348 accounts
[INFO]	2019-07-22T13:00:55.801Z	8f9c695d-a5a1-4e22-838c-d40a8178d7a8	Loading Locations...
[INFO]	2019-07-22T13:01:10.525Z	8f9c695d-a5a1-4e22-838c-d40a8178d7a8	Loaded 16 locations
[INFO]	2019-07-22T13:01:10.526Z	8f9c695d-a5a1-4e22-838c-d40a8178d7a8	Loading Reserved Instances...
[INFO]	2019-07-22T13:01:11.256Z	8f9c695d-a5a1-4e22-838c-d40a8178d7a8	Loaded 513 reserved instances
[INFO]	2019-07-22T13:01:11.258Z	8f9c695d-a5a1-4e22-838c-d40a8178d7a8	Loaded 513 RI Subcriptions
[INFO]	2019-07-22T13:01:11.258Z	8f9c695d-a5a1-4e22-838c-d40a8178d7a8	Loading EC2 Pricing Data...
[INFO]	2019-07-22T13:01:11.310Z	8f9c695d-a5a1-4e22-838c-d40a8178d7a8	Skipping unknown location: AWS GovCloud (US-East)
[INFO]	2019-07-22T13:01:11.311Z	8f9c695d-a5a1-4e22-838c-d40a8178d7a8	Skipping unknown location: EU (Stockholm)
[INFO]	2019-07-22T13:01:11.313Z	8f9c695d-a5a1-4e22-838c-d40a8178d7a8	Skipping unknown location: AWS GovCloud (US)
[INFO]	2019-07-22T13:01:11.318Z	8f9c695d-a5a1-4e22-838c-d40a8178d7a8	Skipping unknown location: Asia Pacific (Osaka-Local)
[INFO]	2019-07-22T13:01:39.253Z	8f9c695d-a5a1-4e22-838c-d40a8178d7a8	Loaded 361625 pricing rows
[INFO]	2019-07-22T13:01:39.282Z	8f9c695d-a5a1-4e22-838c-d40a8178d7a8	Loaded prices for 17320 instance types
[INFO]	2019-07-22T13:01:43.691Z	8f9c695d-a5a1-4e22-838c-d40a8178d7a8	Loaded prices for 65 instance types in ap-east-1
[INFO]	2019-07-22T13:01:43.691Z	8f9c695d-a5a1-4e22-838c-d40a8178d7a8	Loaded prices for 181 instance types in ap-northeast-1
[INFO]	2019-07-22T13:01:43.691Z	8f9c695d-a5a1-4e22-838c-d40a8178d7a8	Loaded prices for 139 instance types in ap-northeast-2
[INFO]	2019-07-22T13:01:43.691Z	8f9c695d-a5a1-4e22-838c-d40a8178d7a8	Loaded prices for 101 instance types in ap-south-1
[INFO]	2019-07-22T13:01:43.691Z	8f9c695d-a5a1-4e22-838c-d40a8178d7a8	Loaded prices for 179 instance types in ap-southeast-1
[INFO]	2019-07-22T13:01:43.691Z	8f9c695d-a5a1-4e22-838c-d40a8178d7a8	Loaded prices for 180 instance types in ap-southeast-2
[INFO]	2019-07-22T13:01:43.691Z	8f9c695d-a5a1-4e22-838c-d40a8178d7a8	Loaded prices for 130 instance types in ca-central-1
[INFO]	2019-07-22T13:01:43.691Z	8f9c695d-a5a1-4e22-838c-d40a8178d7a8	Loaded prices for 170 instance types in eu-central-1
[INFO]	2019-07-22T13:01:43.691Z	8f9c695d-a5a1-4e22-838c-d40a8178d7a8	Loaded prices for 196 instance types in eu-west-1
[INFO]	2019-07-22T13:01:43.692Z	8f9c695d-a5a1-4e22-838c-d40a8178d7a8	Loaded prices for 112 instance types in eu-west-2
[INFO]	2019-07-22T13:01:43.692Z	8f9c695d-a5a1-4e22-838c-d40a8178d7a8	Loaded prices for 104 instance types in eu-west-3
[INFO]	2019-07-22T13:01:43.692Z	8f9c695d-a5a1-4e22-838c-d40a8178d7a8	Loaded prices for 87 instance types in sa-east-1
[INFO]	2019-07-22T13:01:43.692Z	8f9c695d-a5a1-4e22-838c-d40a8178d7a8	Loaded prices for 208 instance types in units
[INFO]	2019-07-22T13:01:43.692Z	8f9c695d-a5a1-4e22-838c-d40a8178d7a8	Loaded prices for 208 instance types in us-east-1
[INFO]	2019-07-22T13:01:43.692Z	8f9c695d-a5a1-4e22-838c-d40a8178d7a8	Loaded prices for 171 instance types in us-east-2
[INFO]	2019-07-22T13:01:43.692Z	8f9c695d-a5a1-4e22-838c-d40a8178d7a8	Loaded prices for 160 instance types in us-west-1
[INFO]	2019-07-22T13:01:43.692Z	8f9c695d-a5a1-4e22-838c-d40a8178d7a8	Loaded prices for 208 instance types in us-west-2
[INFO]	2019-07-22T13:01:43.692Z	8f9c695d-a5a1-4e22-838c-d40a8178d7a8	Querying CUR data from Athena...
```

**Comment by barrymccallarcadia** [comment]
@meckstmd We do utilize i3.metal instances. us-east-1

```
[INFO]	2019-07-22T21:44:58.734Z	e710024c-ecb5-4385-8de7-b48c4cc287e8	Skipping unknown location: Asia Pacific (Osaka-Local)
[INFO]	2019-07-22T21:45:56.475Z	e710024c-ecb5-4385-8de7-b48c4cc287e8	Loaded 105892 pricing rows
[INFO]	2019-07-22T21:45:56.501Z	e710024c-ecb5-4385-8de7-b48c4cc287e8	Loaded prices for xxx instance types
[INFO]	2019-07-22T21:45:59.758Z	e710024c-ecb5-4385-8de7-b48c4cc287e8	Loaded prices for xxx instance types in eu-west-3
[INFO]	2019-07-22T21:45:59.758Z	e710024c-ecb5-4385-8de7-b48c4cc287e8	Loaded prices for xxx instance types in units
[INFO]	2019-07-22T21:45:59.758Z	e710024c-ecb5-4385-8de7-b48c4cc287e8	Loaded prices for xxx instance types in us-east-1
[INFO]	2019-07-22T21:45:59.758Z	e710024c-ecb5-4385-8de7-b48c4cc287e8	Loaded prices for xxx instance types in us-east-2
[INFO]	2019-07-22T21:45:59.758Z	e710024c-ecb5-4385-8de7-b48c4cc287e8	Loaded prices for xxx instance types in us-west-2
[INFO]	2019-07-22T21:45:59.758Z	e710024c-ecb5-4385-8de7-b48c4cc287e8	Querying CUR data from Athena...
[INFO]	2019-07-22T21:46:14.817Z	e710024c-ecb5-4385-8de7-b48c4cc287e8	Loaded 28158 instance summary rows
[INFO]	2019-07-22T21:46:14.819Z	e710024c-ecb5-4385-8de7-b48c4cc287e8	Generating Reports...
'i3.metal': KeyError Traceback (most recent call last): File "/var/task/ariel/lambda.py", line 106, in handler lambda_main(utils.load_config(config)) File "/var/task/ariel/lambda.py", line 42, in lambda_main reports = generate_reports.generate(config, instances, ris, pricing) File "/var/task/ariel/generate_reports.py", line 52, in generate units_value = instances['instancetype'].

**Comment by meckstmd** [comment]
@barrymccallarcadia I have a PR which should fix this issue: https://github.com/yahoo/ariel/pull/4 . Let me know how it works for you.  I can't test it completely for another day since I had to spin up an `i3.metal` instance in one of our accounts since we have none.  I will have to wait for that instance to make it into the Cost and Usage reports before I can fully test this change.  But I was able to verify this change did add it to our ec2 pricing data cache so it should work.

**Comment by meckstmd** [comment]
@barrymccallarcadia Now that we have `i3.metal` instances in our cost and usage report I was able to run Ariel more fully and found another issue to fix so I updated https://github.com/yahoo/ariel/pull/4 with this change.  I could not do a complete test on it yet since our Athena DB is in the UPDATING state.  Tomorrow morning I should be able to do a full run to see if there are any more changes needed.

**Comment by meckstmd** [comment]
@barrymccallarcadia I just merged PR https://github.com/yahoo/ariel/pull/4 so this should now work with your metal instances.  I will leave this issue open to give you a chance to test it.  I also discovered a change in the way AWS exposes CPUCredits which breaks Ariel if you have unlimited `T*` instances with CPU credits so if you have those, try running instead based on PR https://github.com/yahoo/ariel/pull/6

**Comment by meckstmd** [comment]
@barrymccallarcadia I merged PR https://github.com/yahoo/ariel/pull/6 so you should be able to run from the master branch and it should work.

**Comment by barrymccallarcadia** [comment]
@meckstmd compiled up the container for 2.0.3 and encountered another key error: 
```[ERROR]	2019-08-01T20:06:21.82Z	eefd2f27-8f6f-4f01-b245-3d30b905b3a9	Missing RI Pricing data(2) for us-east-1:c1.medium
('Red Hat Enterprise Linux', 'occurred at index (us-east-1, m4, Dedicated, Red Hat Enterprise Linux)'): KeyError
Traceback (most recent call last):
  File "/var/task/ariel/lambda.py", line 106, in handler
    lambda_main(utils.load_config(config))
  File "/var/task/ariel/lambda.py", line 42, in lambda_main
    reports = generate_reports.generate(config, instances, ris, pricing)
  File "/var/task/ariel/generate_reports.py", line 506, in generate
    axis=1)

**Comment by meckstmd** [comment]
@barrymccallarcadia I have been trying to recreate the last error you entered.  I am thinking it has something to do with it having a `Red Hat Enterprise Linux` OS since we do not use that OS.  We have some code to make sure the mapping of the reserved instances to our pricing data lines up: https://github.com/yahoo/ariel/blob/master/ariel/get_reserved_instances.py#L52 . I am wondering if RedHat needs some kind of additional mapping.

Do you have `RedHat Enterprise Linux` reserved instances?  I am thinking based on what I think this may be that you do.  If that's the case, if you can send me your cache file at  `/tmp/cached-reserved-instances.csv` that would help.  I realize that contains private info, but all I really need is the operating system column.  Or if you can just confirm that there is an entry in there that says `Red Hat Enterprise Linux` that might be enough for me.

Also, do you have any more of an error than what you copied? I was able to recreate a similar message by messing with the cache files and it looks like this.  Particularly I am looking for the last few lines.
```
  File ".......lambda.py", line 117, in <module>
    cli()
  File ".......lambda.py", line 114, in cli
    lambda_main(utils.load_config(args.config))
  File ".......lambda.py", line 42, in lambda_main
    reports = generate_reports.generate(config, instances, ris, pricing)
  File "......./generate_reports.py", line 506, in generate
    axis=1)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/pandas/core/frame.py", line 6906, in apply
    return op.get_result()
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/pandas/core/apply.py", line 186, in get_result
    return self.apply_standard()
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/pandas/core/apply.py", line 292, in apply_standard
    self.apply_series_generator()
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/pandas/core/apply.py", line 321, in apply_series_generator
    results[i] = self.f(v)
  File "/Users/mmeckstr/git/ariel/ariel/generate_reports.py", line 505, in <lambda>
    min(x['total_ri_units'], x['total_instance_units']) / get_units(x.name[1] + '.' + reference_sizes[x.name[1]]),
KeyError: ('Red Hat Enterprise Linux', 'occurred at index (us-east-2, c5, Shared, Red Hat Enterprise Linux)')
```

**Comment by barrymccallarcadia** [comment]
@meckstmd sorry, I didn't finish my plate of copy pasta. Yes we do have rhel RIs. 
`File "/var/task/ariel/lambda.py", line 42, in lambda_main
reports = generate_reports.generate(config, instances, ris, pricing)
File "/var/task/ariel/generate_reports.py", line 506, in generate
axis=1)
File "/var/task/pandas/core/frame.py", line 6906, in apply
return op.get_result()
File "/var/task/pandas/core/apply.py", line 186, in get_result
return self.apply_standard()
File "/var/task/pandas/core/apply.py", line 292, in apply_standard
self.apply_series_generator()
File "/var/task/pandas/core/apply.py", line 321, in apply_series_generator
results[i] = self.f(v)
File "/var/task/ariel/generate_reports.py", line 505, in <lambda>
min(x['total_ri_units'], x['total_instance_units']) / get_units(x.name[1] + '.' + reference_sizes[x.name[1]]),
KeyError: ('Red Hat Enterprise Linux', 'occurred at index (us-east-1, m4, Dedicated, Red Hat Enterprise Linux)')`

**Comment by meckstmd** [comment]
@barrymccallarcadia I have a PR out now for you to test to see if this fixes the RHEL issue: https://github.com/yahoo/ariel/pull/7
I can't duplicate this issue in our environment because we do not have any RHEL RIs but I think this PR will fix your issue based on some testing I did with some other RI types we have.

**Comment by meckstmd** [comment]
@barrymccallarcadia I went ahead and merged https://github.com/yahoo/ariel/pull/7 into master so you can do your testing directly on master.  The version for this fix is 2.0.4

**Comment by meckstmd** [comment]
Closing ticket. Feel free to reopen if you are still having this problem or open a new ticket if you experience any other issues.
